src/model/NPMResolver.ts

```diff
@@ -0,0 +1,194 @@
+                    process.exit()
+                } else {
+                    this.checkAndInstall()
+                }
+            }, 500);
+        };
+        nodeModulesWatcher
+            .on('ready', () => ready = true)
+            .on('add', handleEvent)
+            .on('change', handleEvent)
+            .on('unlink', handleEvent)
+            .on('unlinkDir', handleEvent)
+            .on('error', (e) => {
+                console.error(cerr` [nodeModulesWatcher] watcher failed: ${e}`);
+                nodeModulesWatcher.close();
+            });
+    }
+
+    private checkAndInstall() {
+        if (this.hasNoNodeModules()) {
+            console.log(cgreen`⇢` + ` [NPM] There seems to be no installed Node Modules`);
+            this.doNpmInstall();
+        } else {
+            this.checkHashAndInstall();
+        }
+    }
+
+    private checkHashAndInstall() {
+        if (fs.existsSync(this.hashFilePath)) {
+            if (this.packageLockWasUpdated()) {
+                console.log(cgreen`⇢` + ` [NPM] package-lock.json was updated`);
+                this.doNpmInstall();
+            }
+        } else {
+            this.createHashFile();
+            this.doNpmInstall();
+        }
+    }
+
+    private packageLockWasUpdated(): boolean {
+        const oldHash = fs.readFileSync(this.hashFilePath, {encoding: 'utf8'});
+        if (oldHash === this.getHash4PackageLock()) {
+            console.log(cgreen`✓` + ` [NPM] node_modules are up to date`);
+            return false;
+        }
+        return true;
+    }
+
+    private doNpmInstall() {
+        console.log(`⟲ [NPM] executing 'npm install'`);
+        const result = spawn.sync('npm', ['install'], {
+            stdio: [process.stdin, process.stdout, process.stderr]
+        });
+        if (result.status !== 0) {
+            console.log(cred`✗` + ` [NPM] npm install ran into: ${result.error} and failed`);
+            throw Error(cred`✗` + ` [NPM] npm install failed...`);
+        }
+        console.log(cgreen`⇢` + ` [NPM] npm install return code: ${result.status}`);
+        this.createHashFile();
+    }
+
+    private getHash4PackageLock(): string {
+        const hash = crypto.createHash('sha256');
+        const data = fs.readFileSync(this.getPackageLockPath());
+        hash.update(data);
+        return hash.digest('hex');
+    }
+
+    private createHashFile() {
+        fs.writeFileSync(this.hashFilePath, this.getHash4PackageLock(), {encoding: 'utf8'});
+    }
+
+    private hasNoNodeModules(): boolean {
+        const nodeModulesPath = this.getNodeModulesPath();
+        if (fs.existsSync(nodeModulesPath)) {
+            const directories = fs.readdirSync(nodeModulesPath)
+                .filter((name) => fs.statSync(path.join(nodeModulesPath, name)).isDirectory());
+            if (directories.length === 0) {
+                return true;
+            }
+        } else {
+            fs.mkdirSync(this.getNodeModulesPath());
```

Comment:
Method name does not hint at "side effect" - do we need to create a directory here? 🤔